pkg/controller/dynamicupgradeableconditionscontroller.go

```diff
@@ -0,0 +1,245 @@
+			return nil, nil
+		}
+
+		rel, err := store.Deployed(ce.GetName())
+		if errors.Is(err, driver.ErrNoDeployedReleases) {
+			c.logger.Info("Cluster Extension not yet deployed - will check again later")
+			continue
+		}
+		if err != nil {
+			c.logger.V(4).Error(err, "Error returning the last deployed release")
+			return nil, err
+		}
+		props, err := propertyListFromPropertiesAnnotation(rel.Chart.Metadata.Annotations["olm.properties"])
+		if err != nil {
+			c.logger.V(4).Error(err, "Error converting olm.properties")
+			return nil, err
+		}
+		for _, p := range props {
+			if p.Type == MaxOpenShiftVersionProperty {
+				version := string(p.Value)
+				maxOCPVersion, err := toSemver(version)
+				if err != nil {
+					c.logger.V(4).Info(fmt.Sprintf("Error converting to semver for version %s: %v", version, err))
+					continue
+				}
+				if maxOCPVersion != nil && !maxOCPVersion.GTE(next) {
+					// Incompatible
+					incompatibleOperators = append(incompatibleOperators, rel.Labels[BundleNameKey])
+				}
+			}
+		}
+	}
+
+	// deterministic ordering
+	sort.Strings(incompatibleOperators)
+
+	updateStatusFuncs := make([]v1helpers.UpdateStatusFunc, 0, len(incompatibleOperators))
+	if len(incompatibleOperators) > 0 {
+		updateStatusFuncs = append(updateStatusFuncs, v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+			Type:    "Upgradeable",
+			Status:  operatorv1.ConditionFalse,
+			Reason:  incompatibleOperatorsInstalled,
+			Message: strings.Join(incompatibleOperators, ","),
+		}))
+	} else {
+		updateStatusFuncs = append(updateStatusFuncs, v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+			Type:   "Upgradeable",
+			Status: operatorv1.ConditionTrue,
+		}))
+	}
+	return updateStatusFuncs, nil
+}
+
+func propertyListFromPropertiesAnnotation(raw string) ([]property.Property, error) {
+	var props []property.Property
+	if err := json.Unmarshal([]byte(raw), &props); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal properties annotation: %w", err)
+	}
+	return props, nil
+}
+
+func (c *dynamicUpgradeableConditionController) buildHelmStore(secretClient v1.SecretInterface) helm.Storage {
+	log := func(s string, args ...interface{}) { c.logger.V(4).Info(fmt.Sprintf(s, args...)) }
+	csConfig := storage.ChunkedSecretsConfig{
+		ChunkSize:      0,
+		MaxReadChunks:  0,
+		MaxWriteChunks: 0,
+		Log:            log,
+	}
+
+	return helm.Storage{
+		Driver:     storage.NewChunkedSecrets(secretClient, "operator-controller", csConfig),
+		MaxHistory: 0,
+		Log:        log,
+	}
+}
+
+const (
+	releaseEnvVar = "RELEASE_VERSION" // OpenShift's env variable for defining the current release
+)
+
+func nextY(v semver.Version) (semver.Version, error) {
+	v.Build = nil                // Builds are irrelevant
+	v.Pre = nil                  // Next Y release
+	return v, v.IncrementMinor() // Sets Y=Y+1 and Z=0
+}
+
+func toSemver(max string) (*semver.Version, error) {
+	value := strings.Trim(max, "\"")
```

Comment:
`ParseTolerant` does this

Reply:
`ParseTolerant` accepts _more_ than just that though. It accepts 3 and 4 as well. Assuming 2 dominates usage in the catalogs, I think we should reject 1, 3, and 4, and then proactively tell authors using 1, 3, and 4 that they need to switch to 2.

Like I said before, it would be good to have some actual data here. But I don't think it helps anyone for there to be 4 ways to represent the same thing. That just leads to confusion and bugs.

Reply:
I just searched the existing catalogs and found only 1 and 2. Let's reject 3 and 4 for now and then come back to the question of whether to also drop 1 or 2?